File: agent/macos/Sources/FenrirAgent/UpdaterService.swift
import Foundation

private let currentVersion = "1.0.0"
private let githubReleaseAPI = "https://api.github.com/repos/AntiVirus/fenrir/releases/latest"

struct GithubRelease: Codable {
    let tagName: String
    let assets: [Asset]

    struct Asset: Codable {
        let name: String
        let browserDownloadUrl: String
    }

    enum CodingKeys: String, CodingKey {
        case tagName = "tag_name"
        case assets
    }
}

class UpdaterService {
    private let session: URLSession

    init() {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 15
        self.session = URLSession(configuration: cfg)
    }

    func checkForUpdate(completion: @escaping (GithubRelease?, Bool, Error?) -> Void) {
        guard let url = URL(string: githubReleaseAPI) else { return }
        var req = URLRequest(url: url)
        req.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")

        session.dataTask(with: req) { data, _, error in
            if let error = error { completion(nil, false, error); return }
            guard let data = data,
                  let release = try? JSONDecoder().decode(GithubRelease.self, from: data) else {
                completion(nil, false, nil); return
            }
            let latest = release.tagName.trimmingCharacters(in: CharacterSet(charactersIn: "v"))
            let hasUpdate = latest != currentVersion
            if hasUpdate {
                print("[Updater] New version available: \(release.tagName) (current: \(currentVersion))")
            }
            completion(release, hasUpdate, nil)
        }.resume()
    }

    func downloadAndApply(release: GithubRelease, completion: @escaping (Bool, Error?) -> Void) {
        let arch = ProcessInfo.processInfo.environment["ARCH"] ?? "arm64"
        let assetName = "fenrir-macos-agent-\(arch)"
        guard let asset = release.assets.first(where: { $0.name == assetName }),
              let downloadURL = URL(string: asset.browserDownloadUrl) else {
            print("[Updater] No matching asset for \(assetName)")
            completion(false, nil)
            return
        }

        let tmpURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("fenrir-update")
        session.downloadTask(with: downloadURL) { localURL, _, error in
            if let error = error { completion(false, error); return }
            guard let localURL = localURL else { completion(false, nil); return }

            do {
                if FileManager.default.fileExists(atPath: tmpURL.path) {
                    try FileManager.default.removeItem(at: tmpURL)
                }
                try FileManager.default.moveItem(at: localURL, to: tmpURL)
                try FileManager.default.setAttributes([.posixPermissions: 0o755], ofItemAtPath: tmpURL.path)

                // Restart via launchctl
                let selfPath = Bundle.main.executablePath ?? ""
                try FileManager.default.replaceItem(at: URL(fileURLWithPath: selfPath), withItemAt: tmpURL, backupItemName: nil, options: [], resultingItemURL: nil)

                print("[Updater] Updated to \(release.tagName). Restarting...")
                Process.launchedProcess(launchPath: "/bin/launchctl", arguments: ["kickstart", "-k", "system/com.fenrir.agent"])
                completion(true, nil)
            } catch {
                completion(false, error)
            }
        }.resume()
    }
}

File: agent/linux/pkg/updater/UpdaterService.go
package updater

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/exec"
	"runtime"
	"time"
)

const (
	currentVersion   = "1.0.0"
	githubReleaseAPI = "https://api.github.com/repos/AntiVirus/fenrir/releases/latest"
)

type GithubRelease struct {
	TagName string `json:"tag_name"`
	Assets  []struct {
		Name               string `json:"name"`
		BrowserDownloadURL string `json:"browser_download_url"`
	} `json:"assets"`
}

type UpdaterService struct {
	httpClient *http.Client
}

func NewUpdaterService() *UpdaterService {
	return &UpdaterService{
		httpClient: &http.Client{Timeout: 15 * time.Second},
	}
}

// CheckForUpdate queries GitHub for the latest release.
func (u *UpdaterService) CheckForUpdate() (*GithubRelease, bool, error) {
	req, _ := http.NewRequest("GET", githubReleaseAPI, nil)
	req.Header.Set("Accept", "application/vnd.github+json")

	resp, err := u.httpClient.Do(req)
	if err != nil {
		return nil, false, fmt.Errorf("github api: %w", err)
	}
	defer resp.Body.Close()

	var release GithubRelease
	if err := json.NewDecoder(resp.Body).Decode(&release); err != nil {
		return nil, false, fmt.Errorf("decode release: %w", err)
	}

	latest := release.TagName
	if latest == "v"+currentVersion || latest == currentVersion {
		return &release, false, nil
	}

	log.Printf("[Updater] New version available: %s (current: %s)", latest, currentVersion)
	return &release, true, nil
}

// DownloadAndApply downloads the matching asset and hot-swaps the binary.
func (u *UpdaterService) DownloadAndApply(release *GithubRelease) error {
	assetName := fmt.Sprintf("fenrir-linux-agent-%s", runtime.GOARCH)
	var downloadURL string
	for _, asset := range release.Assets {
		if asset.Name == assetName {
			downloadURL = asset.BrowserDownloadURL
			break
		}
	}
	if downloadURL == "" {
		return fmt.Errorf("no matching asset found for %s", assetName)
	}

	tmpPath := "/tmp/fenrir-update"
	if err := u.download(downloadURL, tmpPath); err != nil {
		return fmt.Errorf("download: %w", err)
	}
	if err := os.Chmod(tmpPath, 0755); err != nil {
		return err
	}

	selfPath, _ := os.Executable()
	backupPath := selfPath + ".bak"

	if err := os.Rename(selfPath, backupPath); err != nil {
		return fmt.Errorf("backup self: %w", err)
	}
	if err := os.Rename(tmpPath, selfPath); err != nil {
		os.Rename(backupPath, selfPath)
		return fmt.Errorf("swap binary: %w", err)
	}

	log.Printf("[Updater] Updated to %s. Restarting via systemctl...", release.TagName)
	exec.Command("systemctl", "restart", "fenrir").Start()
	return nil
}

func (u *UpdaterService) download(url, dest string) error {
	resp, err := u.httpClient.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	f, err := os.Create(dest)
	if err != nil {
		return err
	}
	defer f.Close()
	_, err = io.Copy(f, resp.Body)
	return err
}
